### PR TITLE
ND convolution GPU support

### PR DESCRIPTION
**Pull request by kahmed10**
Currently no unit tests, but manually tested for correctness by using conv_1d_test.onnx and conv_3d_test.onnx.

### CODE REVIEW

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp [hunk 1]
```diff
@@ -75,21 +75,31 @@ inline tensor_descriptor make_tensor(const migraphx::shape& os, bool pack = fals
     return t;
 }
 
+inline void insert_attribute_dim(std::vector<int>& padding,
```
Comment (scxiao) [inline src/targets/gpu/include/migraphx/gpu/miopen.hpp:78]: Maybe we can rename it to insert_attribute_1d_dims

File: src/targets/gpu/convolution.cpp [hunk 2]
```diff
@@ -9,15 +9,38 @@ namespace gpu {
 shape miopen_convolution::compute_shape(const std::vector<shape>& inputs) const
 {
     check_shapes{inputs, *this}.has(4).standard();
-    return op.compute_shape({inputs.at(0), inputs.at(1)});
+    std::vector<shape> conv_inputs(inputs.begin(), inputs.begin() + 2);
+    check_shapes{conv_inputs, *this}.max_ndims(5);
+    return op.compute_shape(conv_inputs);
 }
+
+void recompute_shape(shape& input)
```
Comment (scxiao) [inline src/targets/gpu/convolution.cpp:17]: How about the function name extend_1d_shape?

**Comment by kahmed10** [comment]
> We need some 1d/3d unit tests to verify its correctness

We won't be able to add unit tests for correctness until the cpu implementation is finished

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp (hunk 1 above)
Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp]: fixed

File: src/targets/gpu/convolution.cpp (hunk 2 above)
Comment (kahmed10) [reply src/targets/gpu/convolution.cpp]: I changed the name to "recompute_shape_to_2d"

File: src/targets/gpu/convolution.cpp [hunk 3]
```diff
@@ -9,15 +9,38 @@ namespace gpu {
 shape miopen_convolution::compute_shape(const std::vector<shape>& inputs) const
 {
     check_shapes{inputs, *this}.has(4).standard();
-    return op.compute_shape({inputs.at(0), inputs.at(1)});
+    std::vector<shape> conv_inputs(inputs.begin(), inputs.begin() + 2);
+    check_shapes{conv_inputs, *this}.max_ndims(5);
+    return op.compute_shape(conv_inputs);
 }
+
+inline void recompute_shape_to_2d(shape& input)
```
Comment (pfultz2) [inline src/targets/gpu/convolution.cpp:17]: This should return the new shape instead of modifying it.

File: src/targets/gpu/convolution.cpp [hunk 4]
```diff
@@ -9,15 +9,38 @@ namespace gpu {
...
-    auto y_desc = make_tensor(output_shape);
+    shape x_shape = args[0].get_shape();
+    shape w_shape = args[1].get_shape();
+    shape y_shape = output_shape;
+
+    recompute_shape_to_2d(x_shape);
+    recompute_shape_to_2d(w_shape);
+    recompute_shape_to_2d(y_shape);
+
+    auto x_desc = make_tensor(x_shape);
```
Comment (pfultz2) [inline src/targets/gpu/convolution.cpp:41]: These can be written as `auto x_desc = make_tensor(recompute_shape_to_2d(x_shape));`

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp [hunk 5]
```diff
@@ -75,21 +75,31 @@ inline tensor_descriptor make_tensor(const migraphx::shape& os, bool pack = fals
...
         c_mode = miopenGroupConv;
-    miopenInitConvolutionDescriptor(c.get(),
-                                    c_mode,
-                                    op.padding[0],
-                                    op.padding[1],
-                                    op.stride[0],
-                                    op.stride[1],
-                                    op.dilation[0],
-                                    op.dilation[1]);
+    std::vector<int> padding(op.padding.begin(), op.padding.end());
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/miopen.hpp:94]: Padding, stride, and dilation is represented as `std::size_t` not `int`.

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp [hunk 6]
```diff
@@ -75,21 +75,31 @@ inline tensor_descriptor make_tensor(const migraphx::shape& os, bool pack = fals
     return t;
 }
 
+inline void insert_attribute_1d_dims(std::vector<int>& padding,
+                                     std::vector<int>& stride,
+                                     std::vector<int>& dilation)
+{
+    padding.insert(padding.begin(), 0);
+    stride.insert(stride.begin(), 1);
+    dilation.insert(dilation.begin(), 1);
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/miopen.hpp:84]: Can we just default initialize these and then copy backwards into the vector?

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp (hunk 5 above)
Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp:94]: The miopen call below requires the attribute data to be `int*`, so I just copied the values as ints here.

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp (hunk 6 above)
Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp]: changed

File: src/targets/gpu/convolution.cpp (hunk 4 above)
Comment (kahmed10) [reply src/targets/gpu/convolution.cpp]: fixed

File: src/targets/gpu/convolution.cpp (hunk 3 above)
Comment (kahmed10) [reply src/targets/gpu/convolution.cpp]: fixed

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp (hunk 5 above)
Comment (pfultz2) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp]: Ok I see.

File: src/targets/gpu/include/migraphx/gpu/miopen.hpp [hunk 7]
```diff
@@ -91,12 +97,11 @@ inline convolution_descriptor make_conv(const T& op)
...
     if(op.group > 1)
         c_mode = miopenGroupConv;
-    std::vector<int> padding(op.padding.begin(), op.padding.end());
-    std::vector<int> stride(op.stride.begin(), op.stride.end());
-    std::vector<int> dilation(op.dilation.begin(), op.dilation.end());
 
-    if(op.kdims() == 1)
-        insert_attribute_1d_dims(padding, stride, dilation);
+    size_t kdims = op.kdims();
+    std::vector<int> padding = get_attributes(op.padding, kdims, 0);
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/miopen.hpp]: I meant do something like this:

```cpp
std::vector<int> padding(std::min(2, kdims), 0);
std::copy_backwards(op.padding.begin(), op.padding.end(), padding.end());
```

Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp]: this doesn't work if kdims = 3

Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp]: I was able to fix the above by using `std::max` instead

Comment (pfultz2) [reply src/targets/gpu/include/migraphx/gpu/miopen.hpp]: Yea thats right, it should be `std::max`.

**Pull request merged**